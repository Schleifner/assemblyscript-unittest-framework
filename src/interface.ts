/**
 * This file define the interface of coverage information
 */
// input

// instrumented file information
export interface InstrumentResult {
  sourceWasm: string;
  instrumentedWasm: string;
  debugInfo: string;
  expectInfo: string;
}

export type CodeSnippetIndex = number;
export type FunctionIndex = number;
export type LineIndex = number;
export type ColumnIndex = number;
export type FileIndex = number;

// input cov
export type BranchInfo = [CodeSnippetIndex, CodeSnippetIndex];
export type BranchInfos = BranchInfo[];

export type LineRange = [FileIndex, LineIndex, ColumnIndex][];
export type LineInfos = (LineRange | null)[];
export type LineInfoMap = Map<number, Set<LineIndex>>;

export interface CovDebugInfo {
  branchInfo: BranchInfos;
  index: FunctionIndex;
  lineInfo: LineInfos;
}

export interface CovInfo {
  branchInfo: BranchInfos;
  lineInfo: LineInfoMap;
}

export interface DebugInfo {
  debugInfos: Record<string, CovDebugInfo>;
  debugFiles: string[];
}

// input c++ runtime
export type CovTrace = [FunctionIndex, CodeSnippetIndex];
export type TestCaseName = string;

export type ExpectInfoIndex = string;
export type AssertExpectValue = string;
export type AssertActualValue = string;
export type AssertMessages = [ExpectInfoIndex, AssertActualValue, AssertExpectValue][];
export type AssertFailMessage = Record<TestCaseName, AssertMessages>;

export type ErrorMessages = string[];
export type AssertErrorMessages = Map<TestCaseName, ErrorMessages>;

export type ExpectInfoDebugLocation = string;
export type ExpectInfo = Record<ExpectInfoIndex, ExpectInfoDebugLocation>;

export interface IAssertResult {
  fail: number;
  total: number;
  failed_info: AssertFailMessage;
}

// output
export class Rate {
  used = 0;
  total = 0;
  getRate(): number {
    return this.total === 0 ? 100 : Math.round((this.used / this.total) * 10000) / 100;
  }
  toString(): string {
    return `${this.used}/${this.total}`;
  }
  static summarize(rates: Rate[]): Rate {
    return rates.reduce((prev, curr) => {
      prev.used += curr.used;
      prev.total += curr.total;
      return prev;
    }, new Rate());
  }
}
export class FileCoverageResult {
  constructor(public filename: string) {}
  statementCoverageRate: Rate = new Rate();
  branchCoverageRate: Rate = new Rate();
  functionCoverageRate: Rate = new Rate();
  lineCoverageRate: Rate = new Rate();
  sourceUsedCount: CodeCoverage[] = [];
}

export class FunctionCoverageResult {
  constructor(public functionName: string) {}
  branchCoverageRate: Rate = new Rate();
  lineRange: [number, number] = [Number.MAX_SAFE_INTEGER, Number.MIN_SAFE_INTEGER];
  /**
   * first means lineIndex;
   * second means usedCount;
   */
  sourceUsedCount = new Map<number, number>();

  /**
   * Now assemblyscrpt will compile foo<T>() to different function like foo<f64>() , foo<u32>() etc;
   * We need merge the generic function to foo() for coverage statistics
   */
  static mergeFromGeneric(nameWithoutGeneric: string, infos: FunctionCoverageResult[]): FunctionCoverageResult {
    const result = new FunctionCoverageResult(nameWithoutGeneric);
    result.lineRange = [
      Math.min(...infos.map((info) => info.lineRange[0])),
      Math.max(...infos.map((info) => info.lineRange[1])),
    ];
    result.branchCoverageRate = Rate.summarize(infos.map((info) => info.branchCoverageRate));
    for (const info of infos) {
      for (const [lineIndex, count] of info.sourceUsedCount.entries()) {
        const srcLineUsedCount = result.sourceUsedCount.get(lineIndex);
        result.sourceUsedCount.set(lineIndex, srcLineUsedCount === undefined ? count : srcLineUsedCount + count);
      }
    }
    return result;
  }
}

export class CodeCoverage {
  static readonly default = -1;
  source: string;
  /**
   * Default means not a effect line;
   * 0 means not used but effect;
   * more than 0 means a used line;
   */
  usedCount: number = CodeCoverage.default;
  constructor(src: string) {
    this.source = src;
  }
}

export interface UnittestPackage {
  readonly testCodePaths: string[];
  readonly sourceFunctions: Map<string, SourceFunctionInfo[]>;
}

export interface SourceFunctionInfo {
  name: string;
  range: [number, number];
}
