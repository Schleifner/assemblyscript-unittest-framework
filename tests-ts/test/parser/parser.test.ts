import { join } from "node:path";
// eslint-disable-next-line node/no-extraneous-import
import { jest } from "@jest/globals";
import { fileURLToPath, URL } from "node:url";
import { SourceFunctionInfo } from "../../../src/interface.js";

jest.unstable_mockModule("node:fs/promises", () => ({
  readFile: jest.fn(() => "\n".repeat(50)),
}));

// eslint-disable-next-line node/no-unsupported-features/es-syntax
const { Parser } = await import("../../../src/parser/index.js");
// eslint-disable-next-line node/no-unsupported-features/es-syntax
const { readFile } = await import("node:fs/promises");
const dirname = fileURLToPath(new URL(".", import.meta.url));

describe("Parser", () => {
  const parser = new Parser();

  test("traceParse", async () => {
    const debugFilePath = join(dirname, "..", "..", "fixture", "traceParse.debugInfo.json");
    await parser.traceParse(debugFilePath);
    expect(parser.functionCovInfoMap).toMatchSnapshot();
    expect(parser.functionCovTraceMap).toMatchSnapshot();
  });

  test("generateFunctionCoverage", () => {
    parser.generateFunctionCoverage();
    expect(parser.functionCoverageResults.length).toEqual(5);
    expect(parser.functionCoverageResults).toMatchSnapshot();
  });

  test("generateFileCoverage", async () => {
    expect(jest.isMockFunction(readFile)).toBeTruthy();
    const sourceFunction = new Map<string, SourceFunctionInfo[]>();
    sourceFunction.set("test/A.ts", [
      { name: "", range: [39, 39] },
      /** This case means that the line range of this function after instrumented is larger than
       *  the line range generated by transform
       *  It may by caused by inline function or constructor.
       */
      { name: "test/A/Foo#check", range: [20, 24] },
      { name: "", range: [4, 12] },
    ]);
    sourceFunction.set("test/B.ts", [{ name: "test/B/checkMemory", range: [45, 45] }]);
    sourceFunction.set("test/C.ts", [
      { name: "", range: [3, 40] },
      { name: "", range: [42, 44] },
    ]);
    sourceFunction.set("test/D.ts", [{ name: "test/D/visit", range: [10, 11] }]);
    await parser.generateFileCoverage(sourceFunction);
    expect(parser.fileCoverageResults).toMatchSnapshot();
  });
});
