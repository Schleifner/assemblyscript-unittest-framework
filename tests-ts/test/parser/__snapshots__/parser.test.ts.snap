// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser generateFileCoverage 1`] = `
[
  FileCoverageResult {
    "branchCoverageRate": Rate {
      "total": 4,
      "used": 4,
    },
    "filename": "test/A.ts",
    "functionCoverageRate": Rate {
      "total": 3,
      "used": 2,
    },
    "lineCoverageRate": Rate {
      "total": 18,
      "used": 9,
    },
    "sourceUsedCount": [
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": 0,
      },
      CodeCoverage {
        "source": "",
        "usedCount": 0,
      },
      CodeCoverage {
        "source": "",
        "usedCount": 0,
      },
      CodeCoverage {
        "source": "",
        "usedCount": 0,
      },
      CodeCoverage {
        "source": "",
        "usedCount": 0,
      },
      CodeCoverage {
        "source": "",
        "usedCount": 0,
      },
      CodeCoverage {
        "source": "",
        "usedCount": 0,
      },
      CodeCoverage {
        "source": "",
        "usedCount": 0,
      },
      CodeCoverage {
        "source": "",
        "usedCount": 0,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": 3,
      },
      CodeCoverage {
        "source": "",
        "usedCount": 3,
      },
      CodeCoverage {
        "source": "",
        "usedCount": 2,
      },
      CodeCoverage {
        "source": "",
        "usedCount": 2,
      },
      CodeCoverage {
        "source": "",
        "usedCount": 2,
      },
      CodeCoverage {
        "source": "",
        "usedCount": 2,
      },
      CodeCoverage {
        "source": "",
        "usedCount": 1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": 2,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": 3,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
    ],
    "statementCoverageRate": Rate {
      "total": 18,
      "used": 9,
    },
  },
  FileCoverageResult {
    "branchCoverageRate": Rate {
      "total": 1,
      "used": 1,
    },
    "filename": "test/B.ts",
    "functionCoverageRate": Rate {
      "total": 1,
      "used": 1,
    },
    "lineCoverageRate": Rate {
      "total": 1,
      "used": 1,
    },
    "sourceUsedCount": [
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": 4,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
    ],
    "statementCoverageRate": Rate {
      "total": 1,
      "used": 1,
    },
  },
  FileCoverageResult {
    "branchCoverageRate": Rate {
      "total": 4,
      "used": 3,
    },
    "filename": "test/C.ts",
    "functionCoverageRate": Rate {
      "total": 2,
      "used": 1,
    },
    "lineCoverageRate": Rate {
      "total": 13,
      "used": 9,
    },
    "sourceUsedCount": [
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": 2,
      },
      CodeCoverage {
        "source": "",
        "usedCount": 2,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": 4,
      },
      CodeCoverage {
        "source": "",
        "usedCount": 2,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": 2,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": 1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": 1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": 1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": 1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": 0,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": 0,
      },
      CodeCoverage {
        "source": "",
        "usedCount": 0,
      },
      CodeCoverage {
        "source": "",
        "usedCount": 0,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
    ],
    "statementCoverageRate": Rate {
      "total": 13,
      "used": 9,
    },
  },
  FileCoverageResult {
    "branchCoverageRate": Rate {
      "total": 0,
      "used": 0,
    },
    "filename": "test/D.ts",
    "functionCoverageRate": Rate {
      "total": 1,
      "used": 1,
    },
    "lineCoverageRate": Rate {
      "total": 2,
      "used": 2,
    },
    "sourceUsedCount": [
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": 2,
      },
      CodeCoverage {
        "source": "",
        "usedCount": 2,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
      CodeCoverage {
        "source": "",
        "usedCount": -1,
      },
    ],
    "statementCoverageRate": Rate {
      "total": 2,
      "used": 2,
    },
  },
]
`;

exports[`Parser generateFunctionCoverage 1`] = `
[
  FunctionCoverageResult {
    "branchCoverageRate": Rate {
      "total": 0,
      "used": 0,
    },
    "functionName": "test/A/Foo#failString",
    "lineRange": [
      39,
      39,
    ],
    "sourceUsedCount": Map {
      39 => 3,
    },
  },
  FunctionCoverageResult {
    "branchCoverageRate": Rate {
      "total": 4,
      "used": 4,
    },
    "functionName": "test/A/Foo#check",
    "lineRange": [
      16,
      24,
    ],
    "sourceUsedCount": Map {
      16 => 3,
      17 => 3,
      18 => 2,
      19 => 2,
      20 => 2,
      21 => 2,
      22 => 1,
      24 => 2,
    },
  },
  FunctionCoverageResult {
    "branchCoverageRate": Rate {
      "total": 1,
      "used": 1,
    },
    "functionName": "test/B/checkMemory",
    "lineRange": [
      45,
      45,
    ],
    "sourceUsedCount": Map {
      45 => 4,
    },
  },
  FunctionCoverageResult {
    "branchCoverageRate": Rate {
      "total": 4,
      "used": 3,
    },
    "functionName": "test/C/bar",
    "lineRange": [
      10,
      29,
    ],
    "sourceUsedCount": Map {
      10 => 2,
      11 => 2,
      14 => 4,
      15 => 2,
      19 => 2,
      22 => 1,
      24 => 1,
      25 => 1,
      26 => 1,
      29 => 0,
    },
  },
  FunctionCoverageResult {
    "branchCoverageRate": Rate {
      "total": 0,
      "used": 0,
    },
    "functionName": "test/D/visit",
    "lineRange": [
      10,
      11,
    ],
    "sourceUsedCount": Map {
      10 => 2,
      11 => 2,
    },
  },
]
`;

exports[`Parser traceParse 1`] = `
Map {
  "test/A/Foo#failString" => {
    "branchInfo": [],
    "lineInfo": Map {
      0 => Set {
        39,
      },
    },
  },
  "test/A/Foo#emptyFunction" => {
    "branchInfo": [],
    "lineInfo": Map {
      0 => Set {
        56,
      },
      1 => Set {
        57,
      },
    },
  },
  "test/A/Foo#check" => {
    "branchInfo": [
      [
        1,
        2,
      ],
      [
        1,
        3,
      ],
      [
        0,
        1,
      ],
      [
        0,
        5,
      ],
    ],
    "lineInfo": Map {
      0 => Set {
        16,
        17,
      },
      1 => Set {
        18,
        19,
        20,
        21,
      },
      2 => Set {
        22,
      },
      3 => Set {
        24,
      },
      4 => Set {},
      5 => Set {
        24,
      },
    },
  },
  "test/B/checkMemory" => {
    "branchInfo": [
      [
        0,
        1,
      ],
    ],
    "lineInfo": Map {
      0 => Set {
        45,
      },
      1 => Set {
        45,
      },
    },
  },
  "test/C/bar<f32>" => {
    "branchInfo": [
      [
        1,
        2,
      ],
      [
        1,
        3,
      ],
    ],
    "lineInfo": Map {
      0 => Set {
        10,
        11,
      },
      1 => Set {
        14,
      },
      2 => Set {
        15,
      },
      3 => Set {
        19,
      },
    },
  },
  "test/C/bar<~lib/set/Set<f32>>" => {
    "branchInfo": [
      [
        1,
        2,
      ],
      [
        1,
        3,
      ],
    ],
    "lineInfo": Map {
      0 => Set {
        22,
        24,
      },
      1 => Set {
        25,
      },
      2 => Set {
        26,
      },
      3 => Set {
        29,
      },
    },
  },
  "test/D/visit<f32>" => {
    "branchInfo": [],
    "lineInfo": Map {
      0 => Set {
        10,
        11,
      },
    },
  },
}
`;

exports[`Parser traceParse 2`] = `
Map {
  "test/A/Foo#failString" => [
    0,
    0,
    0,
  ],
  "test/A/Foo#check" => [
    0,
    1,
    2,
    0,
    1,
    3,
    0,
    5,
  ],
  "test/B/checkMemory" => [
    0,
    1,
    0,
    1,
  ],
  "test/C/bar<f32>" => [
    0,
    1,
    2,
    1,
    3,
    0,
    1,
    2,
    1,
    3,
  ],
  "test/C/bar<~lib/set/Set<f32>>" => [
    0,
    1,
    2,
  ],
  "test/D/visit<f32>" => [
    0,
    0,
  ],
}
`;
